File: MATLAB/utilities/gc1d_init_oja.m
function [out,params]=gc1d_init_oja(Nmodule, LM_module, LM_phase,onlineplots,lr)

%    Nmodule:number of modules
%    LM_module:LM freq matches which module? (in module#)
%    LM_phase: phase where LM recurrs (in deg)

%set up ec modules
params.vis2mental=500000;  % time points when external LM is off
a = 20;                     % base module frequency
params.lr=lr;
params.a = a;
params.Nmodule=Nmodule;
params.LM_module=LM_module;
params.LM_phase=LM_phase;

params.num_neurons = 360;        % number of neurons in a module
width=params.num_neurons;
totalN = Nmodule*width;

for i = 1:Nmodule
    scale(i) = 1.1^(i-Nmodule/2);
    grid = createbump(width,a*scale(i),1);
    ec(i).module = toeplitz(grid);          % the neurons of a module at all time lags
end
w.ec2lm = randn(width,Nmodule)/totalN;      % initial weight
out.initW = reshape(w.ec2lm,[width Nmodule]);
out.ec = ec;
params.scale=scale;
tti=1;

%%
% lm external activity
idx = LM_phase/360*width;                      % phase of landmark
grid = createbump(width,a,1);
lm.ext = circshift(grid,[idx(1) 1]);

% added by sujay for using multiple landmarks
for ii=2:length(idx)
    lm.ext = lm.ext+circshift(grid,[idx(ii) 1]);

end


% total drive
%%
t = 0;
clf
eta = lr;  %0.0000001;                    % learning rate
while t<1000000                     % simulation duration
    t=t+1;
    lm.int=0;
    for i = 1:Nmodule
        ti(i) = round(t/scale(i));  % equivalent phase step for each module
        thisphase(:,i) = ec(i).module(:,mod(ti(i)-1,width)+1);      % pick the right activity phase
        lm.int = lm.int+w.ec2lm(:,i)'*thisphase(:,i);               % compute internal (EC to LM) input
    end

    Iext = lm.ext(mod(ti(LM_module)-1,width)+1);

    if t<params.vis2mental                 % when external landmark goes off
        lm.s = Iext+lm.int;
    else
        lm.s = lm.int;
    end

    w.ec2lm = w.ec2lm + eta*lm.s*(thisphase-lm.s*w.ec2lm);      % Oja
    w.ec2lm = w.ec2lm - repmat(mean(w.ec2lm),[width 1]);        % remove mean independently for each module

    if mod(t,2000)==0
        t
        if onlineplots==1
            imagesc(reshape(w.ec2lm,[width Nmodule]));hold on
            plot(LM_module,idx,'k.','MarkerSize',20);
            set(gca,'XTick',1:Nmodule,'XTickLabel',round(100*scale)/100);
            xlabel('Spatial scale (a.u.)');
            ylabel('Phase (deg)');
            set(gca,'FontSize',15)
            drawnow
            hold off
        end
        if t<params.vis2mental
            out.vis2mental(tti)=0;
            %                  title 'visual: ext. LM on'
        else
            out.vis2mental(tti)=1;
            %                  title 'mental: ext. LM off'
        end

        out.weights(tti,:,:)=reshape(w.ec2lm,[width Nmodule]);tti=tti+1;
    end

end

end
%%
function bump = createbump(d,width,amp)
z = (-d/2:d/2-1);
kernel = normpdf(z, 0, width);
kernel = amp*kernel/ max(kernel);
bump = kernel';
end

File: Python/utilities/gc1d_init_oja.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm
from scipy.linalg import toeplitz


def createbump(d, width, amp):
    z = np.arange(-d/2, d/2)
    kernel = norm.pdf(z, 0, width)
    kernel = amp * kernel / np.max(kernel)
    return kernel

def gc1d_init_oja(n_module, lm_module, lm_phase, onlineplots, lr):
    class Params:
        pass
    class Out:
        pass
    
    params = Params()
    out = Out()
    
    params.vis2mental = 500000
    a = 20
    params.lr = lr
    params.a = a
    params.n_module = n_module
    params.lm_module = lm_module
    params.lm_phase = lm_phase
    params.num_neurons = 360
    
    width = params.num_neurons
    total_n = n_module * width
    
    scale = np.zeros(n_module)
    ec = []
    for i in range(n_module):
        scale[i] = 1.1**(i + 1 - n_module/2.0)
        grid = createbump(width, a * scale[i], 1)
        ec.append({'module': toeplitz(grid)})
        
    w_ec2lm = np.random.randn(width, n_module) / total_n
    out.init_w = w_ec2lm.copy()
    out.ec = ec
    params.scale = scale
    tti = 0
    
    idx = np.atleast_1d(lm_phase) / 360.0 * width
    grid = createbump(width, a, 1)
    lm_ext = np.roll(grid, int(idx[0]))
    
    for ii in range(1, len(idx)):
        lm_ext = lm_ext + np.roll(grid, int(idx[ii]))
        
    t = 0
    eta = lr
    out.vis2mental_vec = []
    out.weights = []
    
    thisphase = np.zeros((width, n_module))
    ti = np.zeros(n_module, dtype=int)

    while t < 1000000:
        t += 1
        lm_int = 0.0
        for i in range(n_module):
            ti[i] = int(round(t / scale[i]))
            thisphase[:, i] = ec[i]['module'][:, (ti[i] - 1) % width]
            lm_int = lm_int + np.dot(w_ec2lm[:, i], thisphase[:, i])

        i_ext = lm_ext[(ti[lm_module - 1] - 1) % width]

        if t < params.vis2mental:
            lm_s = i_ext + lm_int
        else:
            lm_s = lm_int

        # Oja's rule: w = w + eta * s * (phase - s * w), with lm_s a scalar
        # broadcasting the same update across all modules (columns of w_ec2lm)
        w_ec2lm = w_ec2lm + eta * lm_s * (thisphase - lm_s * w_ec2lm)

        w_ec2lm = w_ec2lm - np.mean(w_ec2lm, axis=0)
        
        if t % 2000 == 0:
            print(t)
            if onlineplots == 1:
                plt.clf()
                plt.imshow(w_ec2lm, aspect='auto')
                plt.plot(lm_module - 1, idx, 'k.', markersize=20)
                plt.xticks(range(n_module), np.round(100 * scale) / 100.0)
                plt.xlabel('Spatial scale (a.u.)')
                plt.ylabel('Phase (deg)')
                plt.pause(0.01)
            
            if t < params.vis2mental:
                out.vis2mental_vec.append(0)
            else:
                out.vis2mental_vec.append(1)
            
            out.weights.append(w_ec2lm.copy())
            tti += 1
            
    out.weights = np.array(out.weights)
    return out, params
